File: Algorithm/2023.12/jonghyo/carry피하기.java
import java.io.*;

public class Main {
	static int answer = 0;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int num[] = new int[N];
		for(int i = 0; i < N; i++) {
			num[i] = Integer.parseInt(br.readLine());
		}
		carry(N, num, 0, 0, 0);
		System.out.println(answer);
	}

	
	
	private static void carry(int n, int[] num, int start, int sum, int count) {
		answer = Math.max(count , answer);
		for(int i = start; i < n; i++) {
			if(check(sum, num[i])) {
				carry(n,num, i+1, sum+num[i], count + 1);
			}
		}
	}

	//체크해줄 수(carry가 발생하는지 체크할 함수입니다.
	private static boolean check(int num1, int num2) {
		int div = 100000000;
		int a = 0;
		int b = 0;
		while(div != 0) {
			a = num1 / div;
			b = num2 / div;
			if(a + b > 9)
				return false;
			num1 %= div;
			num2 %= div;
			div /= 10;
		}
		return true;
	}
}

File: Algorithm/2023.12/wolfy916/연습문제/Carry 피하기.py
'''
Carry 피하기 - 코드트리 실버3
분류 : 백트랙킹
'''
import sys

# [A] 입력함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [B] carry 여부 체크 함수
def check(num1, num2):
    div = 100000000
    while div > 0:
        value1 = num1 // div % 10
        value2 = num2 // div % 10
        if value1 + value2 > 9: return False
        div //= 10
    return True

# [C] 백트랙킹 함수
def dfs(depth, start, value):
    global answer
    if N - start + depth <= answer: return  # 왜 이생각을 못했지
    answer = max(answer, depth)
    for i in range(start, N):
        if check(value, nums[i]):
            dfs(depth + 1, i + 1, value + nums[i])

# [Main]
if __name__ == "__main__":
    N = int(input())
    nums = [int(input()) for _ in range(N)]
    answer = 0
    dfs(0, 0, 0)
    print(answer)